src/compiler/transformers/module/module.ts

```diff
@@ -1315,11 +1315,30 @@ namespace ts {
                 // If we're exporting these variables, then these just become assignments to 'exports.x'.
                 for (const variable of node.declarationList.declarations) {
                     if (isIdentifier(variable.name) && isLocalName(variable.name)) {
+                        // A "local name" generally means a variable declaration that *shouldn't* be
+                        // converted to `exports.x = ...`, even if the declaration is exported. This
+                        // usually indicates a class or function declaration that was converted into
+                        // a variable declaration, as most references to the declaration will remain
+                        // untransformed (i.e., `new C` rather than `new exports.C`). In these cases,
+                        // an `export { x }` declaration will follow.
                         if (!modifiers) {
                             modifiers = visitNodes(node.modifiers, modifierVisitor, isModifier);
                         }
 
-                        variables = append(variables, variable);
+                        if (variable.initializer) {
+                            const updatedVariable = factory.updateVariableDeclaration(
+                                variable,
+                                variable.name,
+                                /*exclamationToken*/ undefined,
+                                /*type*/ undefined,
+                                createExportExpression(

```

Comment:
The new comment and `createExportExpression` appear to conflict: from reading the comment I thought that this block was for declarations that should be exported later, but the new code here calls `createExportExpression` whenever the variable has an initialiser. Why?

Reply:
The comment is unrelated (but still correct). 

When an identifier is a "local name", we don't transform to it into `exports.Foo`. Normally this is because there's an `exports.Foo` "export name" we've defined somewhere else where we do `exports.Foo = Foo`, such as with enums and namespaces below:
```ts
// file.ts
export enum E { ... }

// file.js (--module esnext)
export var E;
(function (E) {
  ...
})(E || (E = {}));

// file.js (--module commonjs)
var E;
//  ^- local name
(function (E) {
  ...
})(E = exports.E || (exports.E = {}));
//     ^^^^^^^^^ - export name
```

Above, the `E` var is uninitialized, so the assignment occurs later. However, with decorated classes this is different:

```ts
// file.ts
export @dec class C {}

// file.js (--module esnext)
export let C = (() => {
  var __classThis;
  var _C_decorators = [dec];
  ...
  var C = class {
    static {
      // decorator emit...
    }
  };
  return C = __classThis;
})();

// file.js (--module commonjs)
var C = exports.C = (() => {
//  ^  ^^^^^^^^^-expression created here
//  local name
  var __classThis;
  var _C_decorators = [dec];
  ...
  var C = class {
    static {
      // decorator emit...
    }
  };
  return C = __classThis;
})();
```

We used to get around this by moving the `export` to a trailing `ExportDeclaration`, i.e.:

```js
// old partial transformation of decorators in the ts transform
@dec
class C {
}
export { C };
```

And that trailing `export { C }` would be turned into `exports.C = C`, but I changed that when I moved more of the decorator-related work out of the `ts` transform, since the result looks cleaner in other module types:

```js
// before (--module esnext)
let C = (() => { /*class transform with decorators*/ })();
export { C };

// before (--module commonjs)
let C = (() => { /*class transform with decorators*/ })();
exports.C = C;

// after (--module esnext)
export let C = (() => { /*class transform with decorators*/ })();

// after(--module commonjs)
let C = exports.C = (() => { /*class transform with decorators*/ })();
```